lib/Tapable.js

```diff
@@ -53,33 +53,37 @@ Tapable.mixin = function mixinTapable(pt) {
 	copyProperties(Tapable.prototype, pt);
 };
 
-Tapable.prototype.applyPlugins = function applyPlugins(name) {
+Tapable.prototype.applyPlugins = function applyPlugins(name, param1, param2) {
 	if(!this._plugins[name]) return;
```

Comment:
I see to performance benefit for the method with variable number of arguments.

But I don't see the benefit for the methods with fixed number of arguments (i. e. `applyPlugins2`). They are manually optimized for a specific number of arguments. Most "hot" hooks use them.

![image](https://user-images.githubusercontent.com/1365881/29325281-6343ce06-81e7-11e7-828e-808e6f0b5e9b.png)

Anyway, I would love to take a more radical approach. This whole package is a bit old. What about a new major version with more performance and more features.

* Never use `arguments`, compile fixed arguments function.
* Create a function per hooks which makes it monomorphic.
* Maybe something magic like ``new Function(`function x(${args}) ...`)``
* Different apply function depending on number of registered plugins
  * no plugins: noop apply for max performance
  * special version to 1 plugin: optimized
  * general version for more plugin
* You need to register hooks before using them. You need to specify number of arguments and kind (sync/async/parallel/waterfall).
* Instead of `compiler.plugin("before-run", ...)` `compiler.hooks.beforeRun(...)` to allow have typings for these hooks.
* Each plugin can have a name `compiler.hooks.beforeRun("HotModuleReplacementPlugin", () => ...)`
* Insert plugins at specify locations instead of always to the end. `hooks.beforeRun({ name: ..., before: "HotModuleReplacementPlugin" }, () => ...)`
* Maybe a middleware api, if this doesn't affect performance: `hooks.beforeRun.intercept(...)`
  * Would be great of the progress display and profiling.
* The `plugin(...)` method can be provided for backward compatibility for plugins.
  * We may generate plugin name from stack trace (depending on performance cost for this)
* The old `applyPluginsXXX` methods can also be offered for backward compatibility, but trigger a deprecation.
* Using a unknown hook name in the `plugin` or `applyPluginsXXX` automatically register a compat hook which handle them.
* The compatibility methods convert dash-case to camelCase.
* This would be a big breaking change, but I guess acceptable with the compatibility layer.